### PR TITLE
Produce errors when empty ids are passed into inspect calls. 

### PR DESCRIPTION
**Pull request by emil2k**
**- What I did**

Made the client validate when empty IDs are passed into inspect methods to prevent calling the wrong endpoint and returning the wrong error.

Previously, if for example a blank `nodeID` was passed in to `NodeInspectWithRaw` the result was a node list request. The response would then fail to unmarshal into a `Node` type returning a JSON error.

**- How I did it**

Checked ids passed into inspect calls and returned an error if they were empty.

**- How to verify it**

Test added.

**- Description for the changelog**
Produce errors when empty ids are passed into inspect calls.

### CODE REVIEW

File: client/node_inspect.go [hunk 1]
```diff
@@ -6,11 +6,15 @@ import (
...
 	"github.com/docker/docker/api/types/swarm"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 )
 
 // NodeInspectWithRaw returns the node information.
 func (cli *Client) NodeInspectWithRaw(ctx context.Context, nodeID string) (swarm.Node, []byte, error) {
+	if nodeID == "" {
+		return swarm.Node{}, nil, errors.New("node id is required")
+	}
```
Comment (thaJeztah) [inline client/node_inspect.go]: I'm not sure this is needed. Other object types don't have this check (e.g. https://github.com/moby/moby/blob/81bb9978ab5ac99e84a5bf62d0d469f0aec1d506/client/service_inspect.go#L15-L34)

There's already validation in place on the CLI;

```bash
$ docker node inspect
"docker node inspect" requires at least 1 argument.
```

And the API server performs additional validation;

```bash
$ curl --unix-socket /var/run/docker.sock http://localhost/nodes/ 
{"message":"page not found"}
```

Not entirely against, but I want to guard against having validation in the client where it's not needed, and try to keep as much as possible on the daemon side

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: client/node_inspect.go (hunk 1 above)
Comment (dnephin) [reply client/node_inspect.go]: Unfortunately I think this is necessary. I added a similar check for volumes in #34770.

The problem is in building the url. If the arg is blank, the url path ends up being `/nodes`, which is a valid URL (https://docs.docker.com/engine/api/v1.30/#operation/NodeList). I'm not sure why that curl failed.

I wish there was a better way of handling this, but I haven't found anything yet.

File: client/node_inspect.go [hunk 2]
```diff
@@ -6,11 +6,15 @@ import (
...
 
 	"github.com/docker/docker/api/types/swarm"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 )
 
 // NodeInspectWithRaw returns the node information.
 func (cli *Client) NodeInspectWithRaw(ctx context.Context, nodeID string) (swarm.Node, []byte, error) {
+	if nodeID == "" {
+		return swarm.Node{}, nil, errors.New("node id is required")
```
Comment (dnephin) [inline client/node_inspect.go]: We should be consistent here, how about an error like this: https://github.com/moby/moby/pull/34770/files#diff-18fbe225ba23b481ee707e859ff03b0cR22

and a test case

File: client/node_inspect.go (hunk 1 above)
Comment (thaJeztah) [reply client/node_inspect.go]: Difference is in the trailing `/`; without a trailing `/`, it's "list nodes", and with a trailing `/`, it's "view/inspect node"

Comment (dnephin) [reply client/node_inspect.go]: Ah, right! So I guess the line below this should be using `path.Join()` to construct the url

File: client/node_inspect.go (hunk 2 above)
Comment (emil2k) [reply client/node_inspect.go]: I updated the error to be consistent with the volume inspect call and added a similar test case.

File: client/node_inspect.go (hunk 1 above)
Comment (emil2k) [reply client/node_inspect.go]: > Difference is in the trailing /; without a trailing /, it's "list nodes", and with a trailing /, it's "view/inspect node"

@thaJeztah My client was making a request for `nodes/` with the trailing `/`, but was receiving the results from the list nodes endpoint, so maybe there is an issue with the routing.

I found a `go-swagger` issue that seems to have removed any meaning associated with a trailing slash explaining this behavior:
https://github.com/go-swagger/go-swagger/pull/906
https://github.com/go-swagger/go-swagger/issues/899#issuecomment-276244306

This could have affected your API server's routing. You might not be seeing this on your local setup because your `Makefile` is using an old version of `go-swagger`:

```
.PHONY: swagger-gen
swagger-gen:
	docker run --rm -v $(PWD):/go/src/github.com/docker/docker \
		-w /go/src/github.com/docker/docker \
		--entrypoint hack/generate-swagger-api.sh \
		-e GOPATH=/go \
		quay.io/goswagger/swagger:0.7.4
```

While the commit that made these changes was added in version 9.0 of `go-swagger`:
https://github.com/go-swagger/go-swagger/commit/5ef60f1b60c740f0b030541c7f2ef329fe2b5a06

Meanwhile, the commit being used in the `Dockerfile` to build Docker is using a commit that points to version 0.13 of `go-wrapper`:
```
# Install go-swagger for validating swagger.yaml
ENV GO_SWAGGER_COMMIT c28258affb0b6251755d92489ef685af8d4ff3eb
RUN git clone https://github.com/go-swagger/go-swagger.git /go/src/github.com/go-swagger/go-swagger \
	&& (cd /go/src/github.com/go-swagger/go-swagger && git checkout -q $GO_SWAGGER_COMMIT) \
	&& go install -v github.com/go-swagger/go-swagger/cmd/swagger
```
https://github.com/go-swagger/go-swagger/commit/c28258affb0b6251755d92489ef685af8d4ff3eb

Here are my client/server versions:

```
Client:
 Version:      17.11.0-ce
 API version:  1.30 (downgraded from 1.34)
 Go version:   go1.9.2
 Git commit:   1caf76c
 Built:        unknown-buildtime
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.4
 Git commit:   402dd4a/17.06.2-ce
 Built:        Fri Nov 10 00:51:08 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by emil2k** [comment]
I think maybe this comment got lost, because it was in a discussion on an old diff:
https://github.com/moby/moby/pull/36144#discussion_r165017263

I'm referring to it here because I think it is relevant to the issue at hand.

**Comment by thaJeztah** [comment]
Thanks @emil2k - regarding the trailing slash; this was the PR I had in mind; https://github.com/moby/moby/pull/24634. I recall a discussion somewhere where it was suggested that having a trailing slash (or not) should route to a different endpoint (but I wasn't really in favour of that). Perhaps it's worth to verify the behaviour of relevant endpoints.

For this change: I discussed this with @dnephin (I thought I left a comment, but apparently forgot 😅), and we agreed that having a validation for _empty_ values in the client does make sense.

If we make this change, I'd prefer it to be done for all similar endpoints (i.e. all endpoints that expect an ID should produce an error indicating that the ID was empty - if it's not currently handled like that)

**Comment by emil2k** [comment]
@thaJeztah I added this check and tests to all the inspect calls.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

@dnephin PTAL

File: client/volume_inspect.go [hunk 3]
```diff
@@ -18,15 +17,12 @@ func (cli *Client) VolumeInspect(ctx context.Context, volumeID string) (types.Vo
...
-	// The empty ID needs to be handled here because with an empty ID the
-	// request url will not contain a trailing / which calls the volume list API
-	// instead of volume inspect
 	if volumeID == "" {
 		return types.Volume{}, nil, objectNotFoundError{object: "volume", id: volumeID}
 	}
 
 	var volume types.Volume
-	resp, err := cli.get(ctx, path.Join("/volumes", volumeID), nil, nil)
+	resp, err := cli.get(ctx, "/volumes/"+volumeID, nil, nil)
```
Comment (dnephin) [inline client/volume_inspect.go]: Why this change? I think using `path.Join()` is correct

Comment (thaJeztah) [reply client/volume_inspect.go]: Looks like we use concatenation in all the other endpoints

Comment (emil2k) [reply client/volume_inspect.go]: I changed it to be consistent with the other similar methods. Also for simplicity, I could not see a case where it would be necessary to use `path.Join` over the simple concat, but I maybe overlooking something.

Comment (dnephin) [reply client/volume_inspect.go]: I think the only case where it would be different is the empty string value. I guess it's fine, we can fix them all at once at some later time.

**Review by dnephin — APPROVED** [review]
LGTM

**Review by tonistiigi — APPROVED** [review]
